Src/Common/CorrelationIdLookupHelper.cs

```diff
@@ -0,0 +1,176 @@
+        /// <returns>true if correlationId was successfully retrieved; false otherwise.</returns>
+        public bool TryGetXComponentCorrelationId(string instrumentationKey, out string correlationId)
+        {
+            if (string.IsNullOrEmpty(instrumentationKey))
+            {
+                // This method cannot throw - it's a Try... method. We cannot proceed any further.
+                correlationId = string.Empty;
+                return false;
+            }
+
+            var found = this.knownCorrelationIds.TryGetValue(instrumentationKey, out correlationId);
+
+            if (found)
+            {
+                return true;
+            }
+            else
+            {
+                // Simplistic cleanup to guard against this becoming a memory hog.
+                if (this.knownCorrelationIds.Keys.Count >= MAXSIZE)
+                {
+                    this.knownCorrelationIds.Clear();
+                }
+
+                try
+                {
+                    // Todo: When this fails, say in the vortex endpoint case, ProfileQueryEndpont is not provided, it may perpetually keep failing.
+                    // We can possibly make it more robust by storing the failure and quiting querying after a few attempts.
+                    // Is that worth the effort?
+
+                    // We wait for <getAppIdTimeout> seconds (which is 0 at this point) to retrieve the appId. If retrieved during that time, we return success setting the correlation id.
+                    // If we are still waiting on the result beyond the timeout - for this particular call we return the failure but queue a task continuation for it to be cached for next time.
+                    Task<string> getAppIdTask = this.provideAppId(instrumentationKey.ToLowerInvariant());
+                    if (getAppIdTask.Wait(GetAppIdTimeout))
+                    {
+                        this.GenerateCorrelationIdAndAddToDictionary(instrumentationKey, getAppIdTask.Result);
+                        correlationId = this.knownCorrelationIds[instrumentationKey];
+                        return true;
+                    }
+                    else
+                    {
+                        getAppIdTask.ContinueWith((appId) =>
+                        {
+                            try
+                            {
+                                this.GenerateCorrelationIdAndAddToDictionary(instrumentationKey, appId.Result);
+                            }
+                            catch (AggregateException ae)
+                            {
+                                CrossComponentCorrelationEventSource.Log.FetchAppIdFailed(ae.Flatten().InnerException.ToInvariantString());
+                            }
+                        });
+
+                        return false;
+                    }
+                }
+                catch (AggregateException ae)
```

Comment:
May it only throw AggregateException from the above code? May concurrent Dictionary operations throw some other type of the exceptions? We might want to catch all here.